crates/common/trie/node_hash.rs

```diff
@@ -48,20 +60,33 @@ impl NodeHash {
     /// The hash will only be considered invalid if it is empty
     /// Aka if it has a default value instead of being a product of hash computation
     pub fn is_valid(&self) -> bool {
-        !matches!(self, NodeHash::Inline(v) if v.is_empty())
+        !matches!(self, NodeHash::Inline(v) if v.1 == 0)
     }
 
     /// Const version of `Default` trait impl
     pub const fn const_default() -> Self {
-        Self::Inline(vec![])
+        Self::Inline(([0; 31], 0))
+    }
+
+    /// Encodes this NodeHash with the given encoder.
+    pub fn encode<'a>(&self, mut encoder: Encoder<'a>) -> Encoder<'a> {
+        match self {
+            NodeHash::Inline(_) => {
+                encoder = encoder.encode_raw(self.as_ref());
+            }
+            NodeHash::Hashed(_) => {
+                encoder = encoder.encode_bytes(self.as_ref());
+            }
+        }
+        encoder
     }
 }
 
 impl From<Vec<u8>> for NodeHash {
     fn from(value: Vec<u8>) -> Self {
         match value.len() {
             32 => NodeHash::Hashed(H256::from_slice(&value)),
-            _ => NodeHash::Inline(value),
```

Comment:
Maybe this one can just use the slice instead of repeating the match and the inline call? Or maybe both can reuse a common function?